### PR TITLE
[Frontend] Add custom label. Fixes #217

### LINKED ISSUE #217: Add a custom label to an operation
Maybe I am mistaken and this is not Kresus' job but I would find great to be able to assign a custom label to an operation.

Since all I see is a number, (even if it is a check number, most of the time I do not remember why I wrote a check. 

If I could say somewhere that this transaction is for example "Library X for book YYY" that would help me 3 months later to know that the check was for the library "X" and for the purpose of a nice book named "YYY".

### PR DESCRIPTION
**Pull request by ZeHiro**

### CODE REVIEW

File: client/components/OperationList.js [hunk 1]
```diff
@@ -165,6 +165,148 @@ function ComputeAttachmentLink(op) {
...
+
+    onBlur(e) {
+        let customLabel = this.dom().value;
+        Actions.SetCustomLabel(this.props.operation, customLabel);
+        // Be optimistic
+        this.props.operation.customLabel = customLabel;
+        this.switchToStaticMode();
+    }
+    
+    onKeyPress(e) {
```
Comment (ZeHiro) [inline client/components/OperationList.js]: There seem to be a bug there. To reproduce: 
- manually sync operations

File: client/backend.js [hunk 2]
```diff
@@ -98,6 +98,10 @@ module.exports = {
     setTypeForOperation: function(operationId, typeId) {
         return this.updateOperation(operationId, {operationTypeID: typeId});
     },
+    
+    setCustomLabel: function(operationId, customLabel) {
+        return this.updateOperation(operationId, {customLabel:customLabel});
```
Comment (bnjbvr) [inline client/backend.js]: nit: `{customLabel}`

File: client/components/OperationList.js [hunk 3]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
     return `operations/${op.id}/`+file;
 }
 
+
+
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: two extra lines

File: client/components/OperationList.js [hunk 4]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+    }
+
+    switchToEditMode() {
+        this.setState({ editMode: true }, function() {
+            this.dom().focus();
+            //We set the cursor at the end
+            try {
+                this.dom().selectionStart = this.props.defaultValue().length;
+            }
+            catch (err){
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: write try/catch statements like this:
```
try {
    // full tab
} catch (e) {
    // full tab
}
```

File: client/components/OperationList.js [hunk 5]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+            editMode: false            
+        };
+    }
+
+    dom() {
+        return this.refs.customlabel.getDOMNode();
+    }
+
+    switchToEditMode() {
+        this.setState({ editMode: true }, function() {
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: use an arrow function to make sure that `this` refers to the right one

File: client/components/OperationList.js [hunk 6]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+            this.dom().focus();
+            //We set the cursor at the end
+            try {
+                this.dom().selectionStart = this.props.defaultValue().length;
```
Comment (bnjbvr) [inline client/components/OperationList.js]: Would this work: `this.dom().selectionStart = this.dom().value.length;` ? In this case, I think you can remove the try/catch

Comment (bnjbvr) [inline client/components/OperationList.js]: Or even `this.dom().selectionStart = (this.dom().value || '').length;`

File: client/components/OperationList.js [hunk 7]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+        }        
+    }
+    
+    render() {
+        if (!this.state.editMode){
+            return (
+                <button
+                    className="form-control text-left btn-transparent"
+                    id={this.props.operation.id}
+                    onClick={this.switchToEditMode.bind(this)}>
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: for the previous 3 lines (that is, inside the <button> tag), can you half-indent (2 spaces), please? It makes a clear visual difference between the attributes and the content.

File: client/components/OperationList.js [hunk 8]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+        return (
+            <input className="form-control"
+                type="text"
+                ref='customlabel'
+                id={this.props.operation.id}
+                defaultValue={this.props.defaultValue()}
+                onBlur={this.onBlur.bind(this)}
+                onKeyUp={this.onKeyUp.bind(this)}
+            />
+            );
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: the indent is wrong here (4 spaces less)

File: client/components/OperationList.js [hunk 9]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
     return `operations/${op.id}/`+file;
 }
 
+
+
+class LabelComponent extends React.Component {
+    constructor(props) {
+        super(props);
```
Comment (bnjbvr) [inline client/components/OperationList.js]: Can you add a few prerequisites on `props`? Like `has(props, 'operation'); has(props, 'defaultValue');`, etc.

File: client/components/OperationList.js [hunk 10]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+}
+
+class OperationListViewLabelComponent extends LabelComponent {
+    constructor(props) {
+        super(props);
+        has(props, 'operation');
+        has(props, 'label');
+        has(props, 'link');
+        this.defaultValue = this.defaultValue.bind(this);
+        this.buttonLabel = this.buttonLabel.bind(this);
```
Comment (bnjbvr) [inline client/components/OperationList.js]: Do we really need these two bindings here?

Comment (bnjbvr) [inline client/components/OperationList.js]: Oops, nevermind my previous comment.

File: client/components/OperationList.js [hunk 11]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+    defaultValue() {
+        if (this.props.operation.customLabel === null || this.props.operation.customLabel.replace(/\ /g, '').length == 0){
+            return this.props.label;
+        }
+        return this.props.operation.customLabel;
+    }
+
+    render() {
+        if (typeof this.props.link !== 'undefined') {
+            return <div className="input-group">{this.props.link}<LabelComponent
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: Can you put the LabelComponent in the next line, and indent consequently, please?

File: client/components/OperationList.js [hunk 12]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+        super(props);
+        has(props, 'operation');
+        has(props, 'label');
+        has(props, 'link');
+        this.defaultValue = this.defaultValue.bind(this);
+        this.buttonLabel = this.buttonLabel.bind(this);
+    }
+
+    buttonLabel() {
+        if (this.props.operation.customLabel === null || this.props.operation.customLabel.replace(/\ /g, '').length == 0){
```
Comment (bnjbvr) [inline client/components/OperationList.js]: `|| this.props.operation.customLabel.trim().length === 0)`

File: client/components/OperationList.js [hunk 13]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+            return <div className="label-button text-uppercase">{this.props.label}</div>;
+        }
+        return <div className="label-button text-uppercase">{this.props.operation.customLabel}</div>;
+    }
+
+    defaultValue() {
+        if (this.props.operation.customLabel === null || this.props.operation.customLabel.replace(/\ /g, '').length == 0){
```
Comment (bnjbvr) [inline client/components/OperationList.js]: ditto

File: client/components/OperationList.js [hunk 14]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+            return this.props.label;
+        }
+        return this.props.operation.customLabel;
+    }
+
+    render() {
+        if (typeof this.props.link !== 'undefined') {
```
Comment (bnjbvr) [inline client/components/OperationList.js]: in the ctor, we have `has(props, 'link');`, which should assert if `props` is `undefined`. Is this dead code?

File: client/components/OperationList.js [hunk 15]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+            return <div className="input-group">{this.props.link}<LabelComponent
+                    operation={this.props.operation}
+                    defaultValue={this.defaultValue}
+                    buttonLabel={this.buttonLabel} /></div>;
+        }
+        return <LabelComponent
+                    operation={this.props.operation}
+                    defaultValue={this.defaultValue}
+                    buttonLabel={this.buttonLabel} />;
```
Comment (bnjbvr) [inline client/components/OperationList.js]: With inheritance, can you instead use `let comp = super.render();` or `let comp = super();`? This should prevent the copying/pasting.

File: client/components/OperationList.js [hunk 16]
```diff
@@ -165,6 +165,142 @@ function ComputeAttachmentLink(op) {
...
+    }
+
+    defaultValue() {
+        if (this.props.operation.customLabel === null || this.props.operation.customLabel.replace(/\ /g, '').length == 0){
+            return this.props.label;
+        }
+        return this.props.operation.customLabel;
+    }
+
+    render() {
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: no need to implement `render()` if it behaves like the parent function.

File: client/css/style.css [hunk 17]
```diff
@@ -265,6 +265,21 @@ body {
    border-style: solid;
 }
 
+.box-transparent {
+    color: inherit;
+    background-color: transparent;
+    border-style: none;
+}
+
+/*Text alignement for button labels*/
```
Comment (bnjbvr) [inline client/css/style.css]: nit: space after *

File: client/css/style.css [hunk 18]
```diff
@@ -265,6 +265,21 @@ body {
...
+.label-button {
+    overflow: hidden;
+    text-overflow: ellipsis;
+    display: block;
+    white-space:nowrap;
```
Comment (bnjbvr) [inline client/css/style.css]: nit: space after :

File: client/store.js [hunk 19]
```diff
@@ -818,6 +831,16 @@ export let Actions = {
...
         });
     },
+    
+    SetCustomLabel(operation, customLabel) {
+        assert(operation instanceof Operation, 'SetCustomLabel expects an Operation as the first argument');
+        assert(typeof customLabel === 'string', 'SetCustomLabel expects a String as second argument');
+        flux.dispatch({
+            type: Events.user.updated_custom_label_of_operation,
+            operation,
+            customLabel: customLabel
```
Comment (bnjbvr) [inline client/store.js]: nit: `customLabel` (it's a shorthand for `customLabel: customLabel`)

File: server/controllers/operations.js [hunk 20]
```diff
@@ -30,9 +30,9 @@ export function preloadOtherOperation(req, res, next, otherOperationID) {
 export async function update(req, res) {
     let attr = req.body;
 
-    // For now, we can only update the category id or operation type of an operation.
-    if (typeof attr.categoryId === 'undefined' && typeof attr.operationTypeID === 'undefined')
-        return sendErr(res, 'missing parameter categoryId or operationTypeID', 400, 'Missing parameter categoryId or operationTypeID');
+    // For now, we can only update the category id, operation type or custom label of an operation.
```
Comment (bnjbvr) [inline server/controllers/operations.js]: We can remove the "For now," as we have so many different fields to update nowadays ;)

File: server/models/operation.js [hunk 21]
```diff
@@ -22,6 +22,7 @@ let Operation = americano.getModel('bankoperation', {
     categoryId: String,
     attachments: Object,         // {linkTranslationKey: String, linkPlainEnglish: String, url: String}
     operationTypeID: String,
+    customLabel: String,
```
Comment (bnjbvr) [inline server/models/operation.js]: Don't forget to also transfer the customLabel on merge! (server side and client side)

**Comment by bnjbvr** [comment]
Thanks for the PR! A lot of minor style comments, but it looks great. I think the operation merging needs to be handled before this PR gets merged (pun unintended). Also, it'd be nice that the free text form field in the search look in the custom label field, but that can be done later, as it's a separate story. Please let me know if you want to address the comments yourself or if I should ;-)

File: client/backend.js (hunk 2 above)
Comment (ZeHiro) [inline client/backend.js]: OK

File: client/components/OperationList.js (hunk 3 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 4 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: ok

File: client/components/OperationList.js (hunk 5 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 6 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 7 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 8 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 9 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 11 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 12 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 13 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/components/OperationList.js (hunk 14 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: shouldn't be ```has(props, 'link');```

File: client/components/OperationList.js (hunk 15 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: didn't get it :S

File: client/components/OperationList.js (hunk 16 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: OK

File: client/css/style.css (hunk 17 above)
Comment (ZeHiro) [inline client/css/style.css]: OK

File: client/css/style.css (hunk 18 above)
Comment (ZeHiro) [inline client/css/style.css]: OK

File: client/store.js (hunk 19 above)
Comment (ZeHiro) [inline client/store.js]: OK

File: server/controllers/operations.js (hunk 20 above)
Comment (ZeHiro) [inline server/controllers/operations.js]: OK

File: server/models/operation.js (hunk 21 above)
Comment (ZeHiro) [inline server/models/operation.js]: Didn't find the merge part client side ...

File: client/components/OperationList.js (hunk 14 above)
Comment (bnjbvr) [inline client/components/OperationList.js]: OK, in this case, you can use the `maybeHas` helper:

`if (maybeHas(this.props, 'link')`

which is equivalent to what you're doing here.

File: server/models/operation.js (hunk 21 above)
Comment (bnjbvr) [inline server/models/operation.js]: I've just checked and the client side shouldn't need any special attention, only the server side will (server/controllers/operations.js)

File: client/components/OperationList.js (hunk 14 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: maybeHas won't help, as props will have 'link' property, but this.props.link could be null (depends if there is a linked file or not.

File: client/components/OperationList.js [hunk 22]
```diff
@@ -165,6 +165,138 @@ function ComputeAttachmentLink(op) {
...
+    switchToEditMode() {
+        this.setState({ editMode: true }, () => {
+            this.dom().focus();
+            //We set the cursor at the end
+            this.dom().selectionStart = (this.dom().value || '').length;
+            /*try {
+                this.dom().selectionStart = this.props.defaultValue().length;
+            } catch (err){
+                this.dom().selectionStart = 0;
+            }*/
```
Comment (bnjbvr) [inline client/components/OperationList.js]: nit: commented code

File: client/components/OperationList.js (hunk 14 above)
Comment (bnjbvr) [inline client/components/OperationList.js]: Alright then, I trust you to do the right test here :-)

File: client/components/OperationList.js (hunk 22 above)
Comment (ZeHiro) [inline client/components/OperationList.js]: Yes, had noticed, but didn'y update ...

**Comment by ZeHiro** [comment]
This works, but I am not realy sure if it should...

**Comment by ZeHiro** [comment]
OK. It seems that adding props in the constructor does not work after all. (The "added props" are deleted at the end)

File: client/components/OperationList.js [hunk 23]
```diff
@@ -149,6 +148,115 @@ function ComputeAttachmentLink(op) {
...
+        return <div className="label-button">{this.props.operation.customLabel}</div>;
+    }
+}
+
+class OperationListViewLabelComponent extends LabelComponent {
+    constructor(props) {
+        has(props, 'operation');
+        has(props, 'link');
+        has(props, 'label');
+        props.buttonLabel = (() => {return this.buttonLabel()});
```
Comment (ZeHiro) [inline client/components/OperationList.js]: it seems props are read only (or modifications are deleted when calling the parent constructor). An 'elegant' solution could be: 
``` javascript
super({
operation: props.operation,
buttonLabel: (()=> { return this.buttonLabel()})});
```

Comment (bnjbvr) [inline client/components/OperationList.js]: OK, in this case I think the elegant way is just to not have an inheritance model at all:

```
class OperationListViewLabelComponent extends ReactComponent {
  render() {
    return <LabelComponent />; // etc
  }
}
```

Does that make sense?

Comment (ZeHiro) [inline client/components/OperationList.js]: Yes, makes sense.

**Comment by bnjbvr** [comment]
Merged locally, thanks for the PR!

**Comment by ZeHiro** [comment]
You're welcome. Can't wait to use it in production!

**Pull request merged**